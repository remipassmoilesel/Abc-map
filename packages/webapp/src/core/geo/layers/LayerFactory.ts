/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import TileLayer from 'ol/layer/Tile';
import { OSM, TileWMS, WMTS, XYZ } from 'ol/source';
import uuid from 'uuid-random';
import VectorSource from 'ol/source/Vector';
import { tileLoadingAuthenticated } from '../tiles/tileLoadingAuthenticated';
import { LayerType, Logger, PredefinedLayerModel, PredefinedMetadata, VectorMetadata, WmsMetadata, WmtsMetadata, XyzMetadata } from '@abc-map/shared';
import { LayerWrapper, PredefinedLayerWrapper, VectorLayerWrapper, WmsLayerWrapper, WmtsLayerWrapper, XyzLayerWrapper } from './LayerWrapper';
import VectorImageLayer from 'ol/layer/VectorImage';
import Geometry from 'ol/geom/Geometry';
import TileSource from 'ol/source/Tile';
import { styleFunction } from '../styles/style-function';
import { tileLoadingPublic } from '../tiles/tileLoadingPublic';
import { WmsSettings, WmtsSettings, XyzSettings } from './LayerFactory.types';
import { prefixedTranslation } from '../../../i18n/i18n';
import { DefaultStyleOptions } from '../styles/StyleFactoryOptions';
import { XyzLayerProperties } from '@abc-map/shared';

const t = prefixedTranslation('LayerFactory:');

const logger = Logger.get('LayerFactory.ts');

export class LayerFactory {
  public static newPredefinedLayer(model: PredefinedLayerModel): PredefinedLayerWrapper {
    let name: string;
    let source: TileSource;
    switch (model) {
      case PredefinedLayerModel.OSM:
        source = new OSM({ tileLoadFunction: tileLoadingPublic() });
        name = t('OpenStreetMap');
        break;
      default:
        throw new Error(`Unhandled predefined layer type: ${model}`);
    }

    const olLayer = new TileLayer({ source });

    const metadata: PredefinedMetadata = {
      id: uuid(),
      name,
      type: LayerType.Predefined,
      active: false,
      opacity: 1,
      visible: true,
      model,
    };

    return LayerWrapper.from<TileLayer<TileSource>, TileSource, PredefinedMetadata>(olLayer).setMetadata(metadata);
  }

  public static newVectorLayer(source?: VectorSource<Geometry>): VectorLayerWrapper {
    const _source = source || new VectorSource();
    const vectorImageLayer = new VectorImageLayer({ style: (f) => styleFunction(DefaultStyleOptions, f), source: _source });

    const metadata: VectorMetadata = {
      id: uuid(),
      name: t('Geometries'),
      type: LayerType.Vector,
      active: false,
      opacity: 1,
      visible: true,
    };

    return LayerWrapper.from<VectorImageLayer<VectorSource<Geometry>>, VectorSource<Geometry>, VectorMetadata>(vectorImageLayer).setMetadata(metadata);
  }

  public static newWmsLayer(settings: WmsSettings): WmsLayerWrapper {
    const tileLoadFunction = settings.auth?.username && settings.auth?.password ? tileLoadingAuthenticated(settings.auth) : tileLoadingPublic();

    const olLayer = new TileLayer({
      extent: settings.extent,
      source: new TileWMS({
        urls: settings.remoteUrls,
        params: { LAYERS: settings.remoteLayerName, TILED: true },
        tileLoadFunction,
        projection: settings.projection?.name,
      }),
    });

    const metadata: WmsMetadata = {
      id: uuid(),
      name: settings.remoteLayerName,
      type: LayerType.Wms,
      active: false,
      opacity: 1,
      visible: true,
      remoteUrls: settings.remoteUrls,
      remoteLayerName: settings.remoteLayerName,
      projection: settings.projection,
      extent: settings.extent,
      auth: settings.auth,
    };

    return LayerWrapper.from<TileLayer<TileSource>, TileWMS, WmsMetadata>(olLayer).setMetadata(metadata);
  }

  public static newWmtsLayer(settings: WmtsSettings): WmtsLayerWrapper {
    if (!settings.sourceOptions) {
      throw new Error('Source options are mandatory');
    }

    const tileLoadFunction = settings.auth?.username && settings.auth?.password ? tileLoadingAuthenticated(settings.auth) : tileLoadingPublic();

    // We do not set extent here as it can have weird behaviors
    const olLayer = new TileLayer({
      source: new WMTS({
        ...settings.sourceOptions,
        tileLoadFunction,
      }),
    });

    const metadata: WmtsMetadata = {
      id: uuid(),
      name: settings.remoteLayerName,
      type: LayerType.Wmts,
      active: false,
      opacity: 1,
      visible: true,
      capabilitiesUrl: settings.capabilitiesUrl,
      remoteLayerName: settings.remoteLayerName,
      auth: settings.auth,
    };

    return LayerWrapper.from<TileLayer<TileSource>, WMTS, WmtsMetadata>(olLayer).setMetadata(metadata);
  }

  public static newXyzLayer(settings: XyzSettings): XyzLayerWrapper {
    const source = new XYZ({
      url: settings.url,
      projection: settings.projection?.name,
      tileLoadFunction: tileLoadingPublic(),
      minZoom: settings.minZoom,
      maxZoom: settings.maxZoom,
    });
    const olLayer = new TileLayer({ source });
    olLayer.set(XyzLayerProperties.MinZoom, settings.minZoom);
    olLayer.set(XyzLayerProperties.MaxZoom, settings.maxZoom);

    const metadata: XyzMetadata = {
      id: uuid(),
      name: t('XYZ_layer'),
      type: LayerType.Xyz,
      active: false,
      opacity: 1,
      visible: true,
      remoteUrl: settings.url,
      projection: settings.projection,
      minZoom: settings.minZoom,
      maxZoom: settings.maxZoom,
    };

    return LayerWrapper.from<TileLayer<TileSource>, XYZ, XyzMetadata>(olLayer).setMetadata(metadata);
  }
}
