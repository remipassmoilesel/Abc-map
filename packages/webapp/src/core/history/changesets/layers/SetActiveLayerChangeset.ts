/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Changeset } from '../../Changeset';
import { MapWrapper } from '../../../geo/map/MapWrapper';
import { LayerWrapper } from '../../../geo/layers/LayerWrapper';
import { getServices } from '../../../Services';

export class SetActiveLayerChangeset extends Changeset {
  public static create(layer: LayerWrapper): SetActiveLayerChangeset {
    const { geo } = getServices();
    const map = geo.getMainMap();
    const previous = geo.getMainMap().getActiveLayer();
    return new SetActiveLayerChangeset(map, previous, layer);
  }

  constructor(private map: MapWrapper, private previous: LayerWrapper | undefined, private next: LayerWrapper) {
    super();
  }

  public async undo(): Promise<void> {
    if (this.previous) {
      this.map.setActiveLayer(this.previous);
    }
  }

  public async execute(): Promise<void> {
    this.map.setActiveLayer(this.next);
  }
}
