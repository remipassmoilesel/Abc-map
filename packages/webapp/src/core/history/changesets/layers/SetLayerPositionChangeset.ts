/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Changeset } from '../../Changeset';
import { MapWrapper } from '../../../geo/map/MapWrapper';
import { LayerWrapper } from '../../../geo/layers/LayerWrapper';
import { getServices } from '../../../Services';

export class SetLayerPositionChangeset extends Changeset {
  public static create(layer: LayerWrapper, nextPosition: number): SetLayerPositionChangeset {
    const { geo } = getServices();
    const map = geo.getMainMap();
    const previousPosition = map.getLayers().findIndex((lay) => lay.getId() === layer.getId());
    if (previousPosition === -1) {
      throw new Error('Layer does not belong to main map');
    }

    return new SetLayerPositionChangeset(map, layer, previousPosition, nextPosition);
  }

  constructor(private map: MapWrapper, private layer: LayerWrapper, private previousPosition: number, private nextPosition: number) {
    super();
  }

  public async undo(): Promise<void> {
    this.map.removeLayer(this.layer);
    this.map.addLayer(this.layer, this.previousPosition);
  }

  public async execute(): Promise<void> {
    this.map.removeLayer(this.layer);
    this.map.addLayer(this.layer, this.nextPosition);
  }
}
