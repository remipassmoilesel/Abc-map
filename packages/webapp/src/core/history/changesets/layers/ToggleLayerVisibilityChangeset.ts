/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Changeset } from '../../Changeset';
import { LayerWrapper } from '../../../geo/layers/LayerWrapper';
import { MapWrapper } from '../../../geo/map/MapWrapper';
import { getServices } from '../../../Services';

export class ToggleLayerVisibilityChangeset extends Changeset {
  public static create(layer: LayerWrapper, state: boolean): ToggleLayerVisibilityChangeset {
    const { geo } = getServices();
    const map = geo.getMainMap();
    return new ToggleLayerVisibilityChangeset(map, layer, state);
  }

  constructor(private map: MapWrapper, private layer: LayerWrapper, private state: boolean) {
    super();
  }

  public async execute(): Promise<void> {
    this.map.setLayerVisible(this.layer, this.state);
  }

  public async undo(): Promise<void> {
    this.map.setLayerVisible(this.layer, !this.state);
  }
}
