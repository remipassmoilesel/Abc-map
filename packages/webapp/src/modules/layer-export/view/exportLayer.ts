/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { VectorLayerWrapper } from '../../../core/geo/layers/LayerWrapper';
import { Format } from './Format';
import { AbcFile, AbcProjection } from '@abc-map/shared';
import { GeoJSON, GPX, KML, WKT } from 'ol/format';

/**
 * Export layer to specified format.
 *
 * You MUST specify the source projection because it is not set on layers by default.
 *
 * @param layer
 * @param sourceProjection
 * @param format
 */
export async function exportLayer(layer: VectorLayerWrapper, sourceProjection: AbcProjection, format: Format): Promise<AbcFile<Blob>[]> {
  const features = layer.getSource().getFeatures();

  // GeoJSON format
  if (Format.GeoJSON === format && layer.isVector()) {
    const geojsonWriter = new GeoJSON();
    const geojson = geojsonWriter.writeFeatures(features, { featureProjection: sourceProjection.name, dataProjection: 'EPSG:4326' });
    return [{ content: new Blob([geojson]), path: layer.getName() + '.geojson' }];
  }

  // GPX format
  else if (Format.GPX === format && layer.isVector()) {
    const gpxWriter = new GPX();
    const gpx = gpxWriter.writeFeatures(features, { featureProjection: sourceProjection.name, dataProjection: 'EPSG:4326' });
    return [{ content: new Blob([gpx]), path: layer.getName() + '.gpx' }];
  }

  // KML format
  else if (Format.KML === format && layer.isVector()) {
    const kmlWriter = new KML();
    const kml = kmlWriter.writeFeatures(features, { featureProjection: sourceProjection.name, dataProjection: 'EPSG:4326' });
    return [{ content: new Blob([kml]), path: layer.getName() + '.kml' }];
  }

  // WKT format
  else if (Format.WKT === format && layer.isVector()) {
    const wktWriter = new WKT();
    const wkt = wktWriter.writeFeatures(features, { featureProjection: sourceProjection.name, dataProjection: 'EPSG:4326' });
    return [{ content: new Blob([wkt]), path: layer.getName() + '.wkt' }];
  }

  // Not supported
  else {
    throw new Error('Unsupported format or layer: format=' + format + ' layer=' + layer.getName());
  }
}
