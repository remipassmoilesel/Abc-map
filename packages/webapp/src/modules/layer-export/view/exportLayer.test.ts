/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { LayerFactory } from '../../../core/geo/layers/LayerFactory';
import Feature from 'ol/Feature';
import { exportLayer } from './exportLayer';
import { Format } from './Format';
import { AbcProjection, BlobIO } from '@abc-map/shared';
import { GeoJSON } from 'ol/format';

describe('exportLayer()', () => {
  const sourceProjection: AbcProjection = { name: 'EPSG:3857' };

  it('GeoJSON', async () => {
    // Prepare
    const layer = LayerFactory.newVectorLayer();
    layer.getSource().addFeatures(testFeatures());

    // Act
    const exported = await exportLayer(layer, sourceProjection, Format.GeoJSON);

    // Assert
    expect(exported).toHaveLength(1);

    const content = await BlobIO.asString(exported[0].content);
    expect(JSON.parse(content)).toMatchSnapshot();
  });

  it('GPX', async () => {
    // Prepare
    const layer = LayerFactory.newVectorLayer();
    layer.getSource().addFeatures(testFeatures());

    // Act
    const exported = await exportLayer(layer, sourceProjection, Format.GPX);

    // Assert
    expect(exported).toHaveLength(1);

    const content = await BlobIO.asString(exported[0].content);
    expect(content).toMatchSnapshot();
  });

  it('KML', async () => {
    // Prepare
    const layer = LayerFactory.newVectorLayer();
    layer.getSource().addFeatures(testFeatures());

    // Act
    const exported = await exportLayer(layer, sourceProjection, Format.KML);

    // Assert
    expect(exported).toHaveLength(1);

    const content = await BlobIO.asString(exported[0].content);
    expect(content).toMatchSnapshot();
  });

  it('WKT', async () => {
    // Prepare
    const layer = LayerFactory.newVectorLayer();
    layer.getSource().addFeatures(testFeatures());

    // Act
    const exported = await exportLayer(layer, sourceProjection, Format.WKT);

    // Assert
    expect(exported).toHaveLength(1);

    const content = await BlobIO.asString(exported[0].content);
    expect(content).toMatchSnapshot();
  });

  /**
   * Return a set of features for testing purposes.
   *
   * We use EPSG:3857 as source projection because it is the default projection used.
   */
  function testFeatures(): Feature[] {
    const geojson = new GeoJSON();
    return geojson.readFeatures(
      {
        type: 'FeatureCollection',
        features: [
          {
            type: 'Feature',
            geometry: {
              type: 'Point',
              coordinates: [-5373603.704493113, 5400461.011186721],
            },
            properties: {
              'abc:style:point:icon': 'twbs/geo-alt-fill.inline.svg',
              'abc:style:point:size': 90,
              'abc:style:point:color': 'rgba(18,90,147,0.9)',
              'abc:feature:selected': false,
            },
            id: 'inzkwakyyX',
          },
          {
            type: 'Feature',
            geometry: {
              type: 'Point',
              coordinates: [-5037755.981993113, 2794282.6845867215],
            },
            properties: {
              'abc:style:point:icon': 'twbs/geo-alt-fill.inline.svg',
              'abc:style:point:size': 90,
              'abc:style:point:color': 'rgba(18,90,147,0.9)',
              'abc:feature:selected': false,
            },
            id: 'Ek3C9txfoI',
          },
          {
            type: 'Feature',
            geometry: {
              type: 'Point',
              coordinates: [-3197310.462693112, -53706.002213278785],
            },
            properties: {
              'abc:style:point:icon': 'twbs/geo-alt-fill.inline.svg',
              'abc:style:point:size': 90,
              'abc:style:point:color': 'rgba(18,90,147,0.9)',
              'abc:feature:selected': false,
            },
            id: 'h5Lft_xAG9',
          },
          {
            type: 'Feature',
            geometry: {
              type: 'LineString',
              coordinates: [
                [-3506290.367393112, 6018420.820586721],
                [-3573459.9118931126, 2740547.048986722],
                [-1733014.3925931118, 362745.17368672136],
              ],
            },
            properties: {
              'abc:style:stroke:color': 'rgba(18,90,147,0.60)',
              'abc:style:stroke:width': 3,
              'abc:feature:selected': false,
            },
            id: 'nzIpj1HQr3',
          },
          {
            type: 'Feature',
            geometry: {
              type: 'LineString',
              coordinates: [
                [-1612109.2124931123, 2989074.3636367214],
                [-1209091.9454931132, 772479.3951367214],
                [-1289695.3988931123, -1309776.4843632793],
              ],
            },
            properties: {
              'abc:style:stroke:color': 'rgba(18,90,147,0.60)',
              'abc:style:stroke:width': 3,
              'abc:feature:selected': false,
            },
            id: '5OdMTSrfAR',
          },
          {
            type: 'Feature',
            geometry: {
              type: 'Polygon',
              coordinates: [
                [
                  [-2995801.8291931134, -2290451.8340632785],
                  [604485.7560068872, -1591888.5712632798],
                  [685089.2094068881, -3365164.5460632797],
                  [-2995801.8291931134, -2290451.8340632785],
                ],
              ],
            },
            properties: {
              'abc:style:stroke:color': 'rgba(18,90,147,0.60)',
              'abc:style:stroke:width': 3,
              'abc:style:fill:color1': 'rgba(18,90,147,0.30)',
              'abc:style:fill:color2': 'rgba(255,255,255,0.60)',
              'abc:style:fill:pattern': 'abc:style:fill:hatching:oblique:right',
              'abc:feature:selected': true,
            },
            id: 'pCEluZ7j--',
          },
        ],
      },
      { dataProjection: sourceProjection.name, featureProjection: sourceProjection.name }
    );
  }
});
