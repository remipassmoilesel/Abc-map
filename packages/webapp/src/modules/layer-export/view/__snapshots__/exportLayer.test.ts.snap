// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exportLayer() GPX 1`] = `"<gpx xmlns=\\"http://www.topografix.com/GPX/1/1\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xsi:schemaLocation=\\"http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd\\" version=\\"1.1\\" creator=\\"OpenLayers\\"><wpt lat=\\"43.57861000039307\\" lon=\\"-48.27190338547443\\"/><wpt lat=\\"24.33499925170213\\" lon=\\"-45.25493196288962\\"/><wpt lat=\\"-0.4824435253768513\\" lon=\\"-28.72192856712481\\"/><rte><rtept lat=\\"47.465691459816924\\" lon=\\"-31.49754227590284\\"/><rtept lat=\\"23.894412236869627\\" lon=\\"-32.10093656041982\\"/><rtept lat=\\"3.256840068241587\\" lon=\\"-15.567933164655011\\"/></rte><rte><rtept lat=\\"25.91917906895911\\" lon=\\"-14.48182345252448\\"/><rtept lat=\\"6.922397550614718\\" lon=\\"-10.86145774542271\\"/><rtept lat=\\"-11.684088217570377\\" lon=\\"-11.585530886843056\\"/></rte></gpx>"`;

exports[`exportLayer() GeoJSON 1`] = `
Object {
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -48.27190338547443,
          43.57861000039307,
        ],
        "type": "Point",
      },
      "id": "inzkwakyyX",
      "properties": Object {
        "abc:feature:selected": false,
        "abc:style:point:color": "rgba(18,90,147,0.9)",
        "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
        "abc:style:point:size": 90,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -45.25493196288962,
          24.33499925170213,
        ],
        "type": "Point",
      },
      "id": "Ek3C9txfoI",
      "properties": Object {
        "abc:feature:selected": false,
        "abc:style:point:color": "rgba(18,90,147,0.9)",
        "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
        "abc:style:point:size": 90,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -28.72192856712481,
          -0.4824435253768513,
        ],
        "type": "Point",
      },
      "id": "h5Lft_xAG9",
      "properties": Object {
        "abc:feature:selected": false,
        "abc:style:point:color": "rgba(18,90,147,0.9)",
        "abc:style:point:icon": "twbs/geo-alt-fill.inline.svg",
        "abc:style:point:size": 90,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            -31.49754227590284,
            47.465691459816924,
          ],
          Array [
            -32.10093656041982,
            23.894412236869627,
          ],
          Array [
            -15.567933164655011,
            3.256840068241587,
          ],
        ],
        "type": "LineString",
      },
      "id": "nzIpj1HQr3",
      "properties": Object {
        "abc:feature:selected": false,
        "abc:style:stroke:color": "rgba(18,90,147,0.60)",
        "abc:style:stroke:width": 3,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            -14.48182345252448,
            25.91917906895911,
          ],
          Array [
            -10.86145774542271,
            6.922397550614718,
          ],
          Array [
            -11.585530886843056,
            -11.684088217570377,
          ],
        ],
        "type": "LineString",
      },
      "id": "5OdMTSrfAR",
      "properties": Object {
        "abc:feature:selected": false,
        "abc:style:stroke:color": "rgba(18,90,147,0.60)",
        "abc:style:stroke:width": 3,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            Array [
              -26.911745713573936,
              -20.146988059052916,
            ],
            Array [
              5.430187936535306,
              -14.153983118218335,
            ],
            Array [
              6.15426107795567,
              -28.917615807384344,
            ],
            Array [
              -26.911745713573936,
              -20.146988059052916,
            ],
          ],
        ],
        "type": "Polygon",
      },
      "id": "pCEluZ7j--",
      "properties": Object {
        "abc:feature:selected": true,
        "abc:style:fill:color1": "rgba(18,90,147,0.30)",
        "abc:style:fill:color2": "rgba(255,255,255,0.60)",
        "abc:style:fill:pattern": "abc:style:fill:hatching:oblique:right",
        "abc:style:stroke:color": "rgba(18,90,147,0.60)",
        "abc:style:stroke:width": 3,
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`exportLayer() KML 1`] = `"<kml xmlns=\\"http://www.opengis.net/kml/2.2\\" xmlns:gx=\\"http://www.google.com/kml/ext/2.2\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xsi:schemaLocation=\\"http://www.opengis.net/kml/2.2 https://developers.google.com/kml/schema/kml22gx.xsd\\"><Document><Placemark id=\\"inzkwakyyX\\"><ExtendedData><Data name=\\"abc:feature:selected\\"><value>false</value></Data><Data name=\\"abc:style:point:color\\"><value>rgba(18,90,147,0.9)</value></Data><Data name=\\"abc:style:point:icon\\"><value>twbs/geo-alt-fill.inline.svg</value></Data><Data name=\\"abc:style:point:size\\"><value>90</value></Data></ExtendedData><Point><coordinates>-48.27190338547443,43.57861000039307</coordinates></Point></Placemark><Placemark id=\\"Ek3C9txfoI\\"><ExtendedData><Data name=\\"abc:feature:selected\\"><value>false</value></Data><Data name=\\"abc:style:point:color\\"><value>rgba(18,90,147,0.9)</value></Data><Data name=\\"abc:style:point:icon\\"><value>twbs/geo-alt-fill.inline.svg</value></Data><Data name=\\"abc:style:point:size\\"><value>90</value></Data></ExtendedData><Point><coordinates>-45.25493196288962,24.33499925170213</coordinates></Point></Placemark><Placemark id=\\"h5Lft_xAG9\\"><ExtendedData><Data name=\\"abc:feature:selected\\"><value>false</value></Data><Data name=\\"abc:style:point:color\\"><value>rgba(18,90,147,0.9)</value></Data><Data name=\\"abc:style:point:icon\\"><value>twbs/geo-alt-fill.inline.svg</value></Data><Data name=\\"abc:style:point:size\\"><value>90</value></Data></ExtendedData><Point><coordinates>-28.72192856712481,-0.4824435253768513</coordinates></Point></Placemark><Placemark id=\\"nzIpj1HQr3\\"><ExtendedData><Data name=\\"abc:feature:selected\\"><value>false</value></Data><Data name=\\"abc:style:stroke:color\\"><value>rgba(18,90,147,0.60)</value></Data><Data name=\\"abc:style:stroke:width\\"><value>3</value></Data></ExtendedData><LineString><coordinates>-31.49754227590284,47.465691459816924 -32.10093656041982,23.894412236869627 -15.567933164655011,3.256840068241587</coordinates></LineString></Placemark><Placemark id=\\"5OdMTSrfAR\\"><ExtendedData><Data name=\\"abc:feature:selected\\"><value>false</value></Data><Data name=\\"abc:style:stroke:color\\"><value>rgba(18,90,147,0.60)</value></Data><Data name=\\"abc:style:stroke:width\\"><value>3</value></Data></ExtendedData><LineString><coordinates>-14.48182345252448,25.91917906895911 -10.86145774542271,6.922397550614718 -11.585530886843056,-11.684088217570377</coordinates></LineString></Placemark><Placemark id=\\"pCEluZ7j--\\"><ExtendedData><Data name=\\"abc:feature:selected\\"><value>true</value></Data><Data name=\\"abc:style:fill:color1\\"><value>rgba(18,90,147,0.30)</value></Data><Data name=\\"abc:style:fill:color2\\"><value>rgba(255,255,255,0.60)</value></Data><Data name=\\"abc:style:fill:pattern\\"><value>abc:style:fill:hatching:oblique:right</value></Data><Data name=\\"abc:style:stroke:color\\"><value>rgba(18,90,147,0.60)</value></Data><Data name=\\"abc:style:stroke:width\\"><value>3</value></Data></ExtendedData><Polygon><outerBoundaryIs><LinearRing><coordinates>-26.911745713573936,-20.146988059052916 5.430187936535306,-14.153983118218335 6.15426107795567,-28.917615807384344 -26.911745713573936,-20.146988059052916</coordinates></LinearRing></outerBoundaryIs></Polygon></Placemark></Document></kml>"`;

exports[`exportLayer() WKT 1`] = `"GEOMETRYCOLLECTION(POINT(-48.27190338547443 43.57861000039307),POINT(-45.25493196288962 24.33499925170213),POINT(-28.72192856712481 -0.4824435253768513),LINESTRING(-31.49754227590284 47.465691459816924,-32.10093656041982 23.894412236869627,-15.567933164655011 3.256840068241587),LINESTRING(-14.48182345252448 25.91917906895911,-10.86145774542271 6.922397550614718,-11.585530886843056 -11.684088217570377),POLYGON((-26.911745713573936 -20.146988059052916,5.430187936535306 -14.153983118218335,6.15426107795567 -28.917615807384344,-26.911745713573936 -20.146988059052916)))"`;
